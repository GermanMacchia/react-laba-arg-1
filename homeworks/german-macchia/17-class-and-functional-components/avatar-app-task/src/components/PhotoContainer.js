import React from "react";
import reloadIcon from "../assets/refresh-reload.svg";

class PhotoContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      image: props.image,
      name: props.name,
      id: props.id,
      setNew: props.setNew,
    };
  }

  reload() {
    this.setState(() => ({ loading: true }));
    fetch("https://tinyfac.es/api/data?limit=1&quality=1")
      .then((res) => res.json())
      .then((json) => {
        this.setState({
          photos: [
            ...this.state.photos,
            {
              id: json[0].id,
              url: json[0].url,
              name: `${json[0].first_name} ${json[0].last_name}`,
            },
          ],
        });
        this.setState(() => ({ loading: false }));
      });
  }

  render() {
    return (
      <div
        onClick={() => this.state.setNew(`From ${this.state.name}`)}
        className="container"
      >
        <img
          className="container__reload_image"
          src={reloadIcon}
          alt="reload"
        />
        <img
          className="container__profile_image"
          src={this.state.image}
          alt={this.state.name}
        />
      </div>
    );
  }
}

export default PhotoContainer;
