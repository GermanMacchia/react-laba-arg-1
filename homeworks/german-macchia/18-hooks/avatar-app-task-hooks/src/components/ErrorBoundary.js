import React, { Component } from "react";

class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  componentDidCatch(error, errorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo,
    });
  }

  render() {
    if (this.state.errorInfo) {
      return (
        <div className="errorBoundary">
          <div className="errorBoundary__info">
            <h2>Something went wrong</h2>
            <details style={{ whiteSpace: "pre-wrap" }}>
              <br />
              {this.state.error && this.state.error.toString()}
            </details>
          </div>
        </div>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
