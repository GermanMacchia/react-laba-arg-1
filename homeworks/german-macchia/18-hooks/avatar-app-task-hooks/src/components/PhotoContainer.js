import React, { useState, useEffect, useRef } from "react";
import fetchPhotos from "../fetchPhotos";

export const PhotoContainer = React.memo(({ url, name, id }) => {
  const [photo, setPhoto] = useState({ url: url, name: name, id: id });
  const [error, setError] = useState({ error: false, errorMessage: "" });
  const counter = useRef(0);

  // reload new data for this single container
  const reload = async () => {
    const cant = 1;
    const response = await fetchPhotos(cant);
    if (response.status !== 200) {
      setError(() => ({
        error: true,
        errorMessage: `STATUS ${response.status}, ${response.statusText}`,
      }));
    } else {
      const [photo] = await response.json();
      setPhoto(() => ({
        id: photo.first_name + photo.id,
        url: photo.url,
        name: `${photo.first_name} ${photo.last_name}`,
      }));
    }
  };


  useEffect(() => {
    if (error.error) {
      throw new Error(error.errorMessage);
    }
  }, [error]);

  return (
    <div
      className="container"
      onClick={() => reload()}
      key={photo.id}
    >
      {console.log("render App: " + counter.current++)}
      <img
        className="container__profile_image"
        src={photo.url}
        alt={photo.name}
      />
      <div className="container__image_box" />
    </div>
  );
})
