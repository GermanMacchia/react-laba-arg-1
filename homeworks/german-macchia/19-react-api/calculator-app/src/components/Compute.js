import React, { useEffect, useState } from "react";

export const Compute = ({ command, press }) => {
  const [resolution, setResolution] = useState("");
  const [calculation, setCalculation] = useState("");

  const isEmpty = (string) => string === "";
  //Delete last value of calculation field
  const backspace = () => {
    const newValue = calculation.slice(0, -1);
    setCalculation(newValue);
  };
  //Clear calculation field, if it's empty, sets resolution field to 0
  const clear = () => {
    if (isEmpty(calculation)) {
      setResolution(0);
    }
    setCalculation("");
  };
  //Sets calculation field ready to compute solution
  const resolve = async () => {
    //Set regex for not numbers and dot
    const regex = /[^0-9.]/g;
    //Split calculation string on each char and return values that are not numbers nor dots (operators)
    const signs = calculation.split("").filter((value) => regex.test(value));
    //split string were there are not numbers or dot
    const nums = calculation.split(regex);

    //If it is allready a previous resolution and first character it's an operator
    //As the nums value will be splited at first with empty value e.g ["", 23]
    //will put the previous resolution as first value for compute
    if (regex.test(calculation.charAt(0))) {
      nums[0] = resolution;
    }
    //After last condition, if there still no value at index 0 (no resolution),
    //will put a 0 to compute operations
    if (isEmpty(nums[0])) {
      nums[0] = 0;
    }
    //If we put an lonely operator will put this values to compute
    if (isEmpty(nums[1])) {
      nums[0] = resolution || 0;
      nums[1] = 1;
    }

    //each value of the calculation field will be processed in order of input
    //(not in mathematical order of operation )
    let j = 0;
    let result;

    for (let i = 0; i <= signs.length; i++) {
      if (result && nums[i]) {
        result = compute(result, signs[j], nums[i]);
        i++;
      } else {
        result = compute(nums[i], signs[j], nums[i + 1]);
        i++;
      }
      j++;
    }

    setResolution(result);
    clear();
  };

  const compute = (first, op, second) => {
    let num1 = parseFloat(first);
    let num2 = parseFloat(second);
    let res;
    switch (op) {
      case "+":
        res = num1 + num2;
        break;
      case "-":
        res = num1 - num2;
        break;
      case "*":
        res = num1 * num2;
        break;
      case "/":
        if (num2 !== 0) {
          res = num1 / num2;
        } else {
          res = "error";
        }
        break;
      default:
        res = num1;
        break;
    }
    return res;
  };

  useEffect(() => {
    //If last char in calculation string & command are a non digit value
    //and also command is NOT a 'equal, delete, or clear' command value
    //Then, replace that last operator command
    if (
      /\D/g.test(calculation.slice(-1)) &&
      /\D/g.test(command) &&
      /[^CD=]/g.test(command)
    ) {
      setCalculation(calculation.slice(0, -1) + command);
      return;
    }

    //Swich command of 'equal, delete, or clear' if exists
    switch (command) {
      case "=":
        resolve();
        return;
      case "D":
        backspace();
        return;
      case "C":
        clear();
        return;
      default:
        break;
    }

    //set max 17 char length in calculation
    if (calculation.length < 17) {
      setCalculation(calculation + command);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [command, press]);

  return (
    <div className="display">
      <div
        className={
          resolution.toString().length > 9
            ? "display__resolution display__resolution--small"
            : "display__resolution"
        }
      >
        <h1>{resolution.toString().slice(0, 12)}</h1>
      </div>
      <div className="display__calculation">
        <h1>{calculation}</h1>
      </div>
    </div>
  );
};
