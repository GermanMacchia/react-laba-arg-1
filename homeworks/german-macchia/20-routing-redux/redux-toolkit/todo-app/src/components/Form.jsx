import { useDispatch } from "react-redux";
import React, { useRef, useState } from "react";
import { addTodo } from "../features/todos/todosSlice";

export const EntryForm = () => {
  const [todo, setTodo] = useState("");
  const dispatch = useDispatch();
  const form = useRef(null);

  const handleChange = (e) => {
    setTodo({
      [e.target.name]: e.target.value,
    });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(addTodo(todo));
    form.current.reset();
  };

  return (
    <form ref={form}>
      <input
        placeholder="Create Todo-Task"
        type="text"
        name="content"
        onChange={handleChange}
      ></input>
      <button onClick={handleSubmit}>Add</button>
    </form>
  );
};
