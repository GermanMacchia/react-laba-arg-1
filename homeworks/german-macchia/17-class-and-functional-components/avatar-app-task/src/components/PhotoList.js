import PhotoContainer from "./PhotoContainer";
import React, { Component } from "react";
import fetchPhotos from "../api";

class PhotoList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      photos: [],
      refreshList: this.refreshList,
    };
  }

  getPhotos = async (cant) => {
    const response = await fetchPhotos(cant);
    if (response.status !== 200) {
      throw new Error("Too many Request, try later");
    } else {
      const json = await response.json();
      return json.map((photo) => {
        return {
          id: photo.first_name + photo.id,
          url: photo.url,
          name: `${photo.first_name} ${photo.last_name}`,
        };
      });
    }
  };

  refreshList = async () => {
    const statePhotosLength = this.state.photos.length;
    const newPhotos = await this.getPhotos(statePhotosLength);
    this.setState({ photos: newPhotos });
  }

  addPhoto = async () => {
    const cant = 1
    const photo = await this.getPhotos(cant)
    this.setState((state) => ({ photos: [...state.photos, ...photo] }));
  }

  displayPhotos = () =>
    this.state.photos.map((photo) => {
      return (
        <PhotoContainer
          url={photo.url}
          name={photo.name}
          key={photo.name + photo.id}
          id={photo.id}
          reload={this.reloadPhoto}
        />
      );
    });

  render() {
    return <>{this.displayPhotos()}</>;
  }
}

export default PhotoList;
