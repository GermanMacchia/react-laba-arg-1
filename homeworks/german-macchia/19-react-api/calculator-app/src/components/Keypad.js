import { Button } from "../components/Button";
import deleteIcon from "../assets/delete_icon.svg";
import divideIcon from "../assets/divide_icon.svg";
import minusIcon from "../assets/minus_icon.svg";
import plusIcon from "../assets/plus_icon.svg";
import xIcon from "../assets/x_icon.svg";
import equalIcon from "../assets/equal_icon.svg";
import React from "react";

export const Keypad = React.memo(({ handleOnPushKey }) => {
  const deleteInnerIcon = <img src={deleteIcon} alt="delete icon" />;
  const divideInnerIcon = <img src={divideIcon} alt="divide icon" />;
  const minusInnerIcon = <img src={minusIcon} alt="dminus icon" />;
  const plusInnerIcon = <img src={plusIcon} alt="plus icon" />;
  const xInnerIcon = <img src={xIcon} alt="x icon" />;
  const equalInnerIcon = <img src={equalIcon} alt="equal icon" />;

  return (
    <div className="button-grid bottom-rounded--all">
      <Button
        boxStyle="button border-box corner--left"
        innerHtml="C"
        value="C"
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button border-box"
        innerHtml={divideInnerIcon}
        value="/"
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button border-box"
        innerHtml={xInnerIcon}
        value="*"
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button border-box corner--right"
        value="D"
        innerHtml={deleteInnerIcon}
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box"
        innerHtml="7"
        value={7}
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box"
        innerHtml="8"
        value={8}
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box"
        innerHtml="9"
        value={9}
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button border-box"
        innerHtml={minusInnerIcon}
        value="-"
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box"
        innerHtml="4"
        value={4}
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box"
        innerHtml="5"
        value={5}
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box"
        innerHtml="6"
        value={6}
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button border-box"
        innerHtml={plusInnerIcon}
        value="+"
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box"
        innerHtml="1"
        value={1}
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box"
        innerHtml="2"
        value={2}
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box"
        innerHtml="3"
        value={3}
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button large-box bottom-rounded--right"
        innerHtml={equalInnerIcon}
        value="="
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box bottom-rounded--left"
        innerHtml="%"
        value="%"
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box"
        innerHtml="0"
        value={0}
        handleOnPushKey={handleOnPushKey}
      />
      <Button
        boxStyle="button middle-box"
        innerHtml="."
        value="."
        handleOnPushKey={handleOnPushKey}
      />
    </div>
  );
});
