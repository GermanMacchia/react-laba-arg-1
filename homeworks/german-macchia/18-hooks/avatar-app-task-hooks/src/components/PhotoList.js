/* eslint-disable react-hooks/exhaustive-deps */
import { PhotoContainer } from "./PhotoContainer";
import ErrorBoundary from "./ErrorBoundary";
import React, { useState, useEffect, useRef } from "react";
import fetchPhotos from "../fetchPhotos";

export default function PhotoList({
  cantPhotos,
  handleLoading,
  refresh,
  handleRefresh,
}) {
  const [data, setData] = useState([]);
  const [error, setError] = useState({ error: false, errorMessage: "" });
  const [photos, setPhotos] = useState([]);

  //call data from api into data array, set 10 values as default
  const init = async () => {
    const initialValue = 10;
    const response = await fetchPhotos(initialValue);
    if (response.status !== 200) {
      setError({
        error: true,
        errorMessage: `STATUS ${response.status}, ${response.statusText}`,
      });
    } else {
      const json = await response.json();
      const avatars = json.map((photo) => {
        return {
          id: photo.first_name + photo.id,
          url: photo.url,
          name: `${photo.first_name} ${photo.last_name}`,
        };
      });
      setData(avatars);
      handleLoading();
    }
  };

  //Push single photo into array from data array.
  //When data array is empty calls API again
  const pushPhoto = () => {
    if (data.length > 0) {
      const auxData = [...data];
      const newPhoto = auxData.pop();
      setPhotos((photos) => [...photos, newPhoto]);
      if (auxData.length === 0) {
        handleLoading();
        init();
      } else {
        setData(auxData);
      }
    }
  };

  //refresh as many values there are in photos
  const refreshAll = async () => {
    const response = await fetchPhotos(cantPhotos);
    if (response.status !== 200) {
      setError({
        error: true,
        errorMessage: `STATUS ${response.status}, ${response.statusText}`,
      });
    } else {
      const json = await response.json();
      const avatars = json.map((photo) => {
        return {
          id: photo.first_name + photo.id,
          url: photo.url,
          name: `${photo.first_name} ${photo.last_name}`,
        };
      });
      setPhotos(avatars);
      handleRefresh();
    }
  };

  //will throw error if error state changes
  useEffect(() => {
    if (error.error) {
      throw new Error(error.errorMessage);
    }
  }, [error]);

  //call once when render
  useEffect(() => {
    init();
  }, []);

  //call when refresh is set into TRUE
  useEffect(() => {
    if (refresh) {
      refreshAll();
    }
  }, [refresh]);

  //call every time when cantPhotos changes
  useEffect(() => {
    if (data) {
      pushPhoto();
    }
  }, [cantPhotos]);

  return (
    <>
      {photos &&
        photos.map((photo) => {
          return (
            <ErrorBoundary key={photo.name + photo.id}>
              <PhotoContainer url={photo.url} name={photo.name} id={photo.id} />
            </ErrorBoundary>
          );
        })}
    </>
  );
}
