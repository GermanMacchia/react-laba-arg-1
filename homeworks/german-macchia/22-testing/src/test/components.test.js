/**
 * @jest-environment jsdom
 */
import { render, screen } from "@testing-library/react";
import { Button } from "../components/Button";

describe("Test in <Button />", () => {

  const props = {
    boxStyle: "button border-box corner--left",
    innerHtml: "AC",
    value: 0,
  };

  const button = (
    <Button
      innerHtml={props.innerHtml}
      boxStyle={props.boxStyle}
      value={props.value}
    />
  );

  test("should be a match between render and snapshot", () => {
    render(button);
    expect(screen.container).toMatchSnapshot();
  });

  test("should match text in innerText data value", () => {
    render(button);
    expect(screen.getByTestId("innerText").innerHTML).toBe(props.innerHtml);
  });
});
