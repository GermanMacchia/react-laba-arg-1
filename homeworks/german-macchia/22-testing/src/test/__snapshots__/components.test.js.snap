// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <Button /> should be a match between render and snapshot 1`] = `undefined`;
